### PR TITLE
Added support to set BlobParallelUploadOptions.computeMd5 so the service can perform an md5 verification.

### PR DESCRIPTION
**Pull request by gapra-msft**
Resolves https://github.com/Azure/azure-sdk-for-java/issues/12291

Added support to set BlobParallelUploadOptions.computeMd5 so the service can perform an md5 verification.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 1]
```diff
@@ -429,14 +429,16 @@ private SpecializedBlobClientBuilder prepareBuilder() {
...
                 uploadInChunks(blockBlobAsyncClient, stream, validatedParallelTransferOptions,
                     options.getHeaders(), options.getMetadata(), options.getTags(),
-                    options.getTier(), validatedRequestConditions);
+                    options.getTier(), validatedRequestConditions, options.isComputeMd5());
 
             BiFunction<Flux<ByteBuffer>, Long, Mono<Response<BlockBlobItem>>> uploadFullBlobMethod =
-                (stream, length) -> blockBlobAsyncClient.uploadWithResponse(new BlockBlobSimpleUploadOptions(
-                    ProgressReporter.addProgressReporting(stream,
-                        validatedParallelTransferOptions.getProgressReceiver()), length)
+                (stream, length) -> UploadUtils.computeMd5(ProgressReporter.addProgressReporting(stream,
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java:435]: This is the only part of the code that I'm iffy about. Like wrapping the ProgressReporter and stuff looks a little unclean. I could just make a method maybe and make it look a little better.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/options/BlobParallelUploadOptions.java [hunk 2]
```diff
@@ -214,6 +215,25 @@ public BlobParallelUploadOptions setRequestConditions(BlobRequestConditions requ
...
 
+    /**
+     * @return Whether or not the library should calculate the md5 and send it for the service to verify.
+     */
+    public boolean isComputeMd5() {
+        return computeMd5;
+    }
+
+    /**
+     * Sets the calculateAndVerifyMd5 property.
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/options/BlobParallelUploadOptions.java:226]: Rename to computeMd5

File: sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/implementation/UploadUtilsTest.groovy [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        where:
+        data             || _
+        ["Hello World!"] || _
+        ["Hello ", "World!"] || _
+        ["H", "e", "l", "l", "o", " ", "W", "o", "r", "l", "d", "!"] || _
+        ["Hel", "lo World!"] || _
+    }
+
+    @Unroll
+    def "computeMd5 data"() {
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/implementation/UploadUtilsTest.groovy:47]: Add comment that this test is for data integrity

**Review by rickle-msft — APPROVED** [review]

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/UploadUtils.java [hunk 4]
```diff
@@ -142,4 +147,48 @@ public static void uploadFileCleanup(AsynchronousFileChannel channel, ClientLogg
...
+        }
+    }
+
+    public static class FluxMd5Wrapper {
+        private final Flux<ByteBuffer> data;
+        private final byte[] md5;
+
+        FluxMd5Wrapper(Flux<ByteBuffer> data, byte[] md5) {
+            this.data = data;
+            this.md5 = md5 == null ? null : md5.clone();
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/UploadUtils.java:183]: `CoreUtils.clone` should handle this.

**Review by alzimmermsft — APPROVED** [review]

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/UploadUtils.java (hunk 4 above)
Comment (gapra-msft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/UploadUtils.java:183]: # @
```suggestion
            this.md5 = CoreUtils.clone(md5);
```

Comment (gapra-msft) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/UploadUtils.java]: fixed.

**Pull request merged**